cpp/src/arrow/compute/kernels/vector_replace_test.cc

```diff
@@ -793,5 +851,772 @@ TYPED_TEST(TestReplaceBinary, ReplaceWithMaskRandom) {
+  this->AssertFillNullArray(FillNullBackward,
+                            this->array(R"(["tyu", "afd", null, null, null, null])"),
+                            this->array(R"(["tyu", "afd" ,null, null, null, null])"));
+  this->AssertFillNullArray(FillNullBackward,
+                            this->array(R"(["tyu", "afd", null, "qwe", null, null])"),
+                            this->array(R"(["tyu", "afd" , "qwe", "qwe", null, null])"));
+  this->AssertFillNullArray(
+      FillNullBackward,
+      this->array(R"(["tyu", "afd", null, "qwe", null, null, "oiutyu"])"),
+      this->array(R"(["tyu", "afd" ,"qwe", "qwe", "oiutyu", "oiutyu", "oiutyu"])"));
+  this->AssertFillNullArray(
+      FillNullBackward, this->array(R"(["tyu", "afd", null, "qwe", null, null, "qwe"])"),
+      this->array(R"(["tyu", "afd" ,"qwe" , "qwe", "qwe", "qwe", "qwe"])"));
+  this->AssertFillNullArray(
+      FillNullBackward,
+      this->array(R"(["tyu", "afd", null, "qwe", null, "oiutyu", null])"),
+      this->array(R"(["tyu", "afd" ,"qwe" , "qwe", "oiutyu", "oiutyu", null])"));
+  this->AssertFillNullArray(
+      FillNullBackward,
+      this->array(R"(["tyu", "afd", null, "qwe", null, "oiutyu", "aaaagggbbb"])"),
+      this->array(R"(["tyu", "afd" ,"qwe", "qwe", "oiutyu", "oiutyu", "aaaagggbbb"])"));
+  this->AssertFillNullArray(
+      FillNullBackward,
+      this->array(R"(["tyu", "afd" ,"qwe", "qwe", "oiutyu", "oiutyu", "aaaagggbbb"])"),
+      this->array(R"(["tyu", "afd" ,"qwe", "qwe", "oiutyu", "oiutyu", "aaaagggbbb"])"));
+}
+
+// For Test Blocks
+TYPED_TEST(TestFillNullNumeric, FillNullForwardLargeInput) {
+  using CType = typename TypeTraits<TypeParam>::CType;
+  random::RandomArrayGenerator rand(/*seed=*/1000);
+  int64_t len_null = 500;
+  int64_t len_random = 1000;
+  std::shared_ptr<Array> array_random =
+      rand.Numeric<TypeParam>(len_random, /*min=*/0, /*max=*/200, /*nulls=*/0);
+
+  if (array_random) {
+    auto x_ptr = array_random->data()->template GetValues<CType>(1);
+    ASSERT_OK_AND_ASSIGN(auto array_null,
+                         MakeArrayOfNull(array_random->type(), len_null));
+    auto array_null_filled =
+        ConstantArrayGenerator::Numeric<TypeParam>(len_null, x_ptr[len_random - 1]);
+    {
+      ASSERT_OK_AND_ASSIGN(auto value_array,
+                           Concatenate({array_random, array_null, array_random}));
+      ASSERT_OK_AND_ASSIGN(auto result_array,
+                           Concatenate({array_random, array_null_filled, array_random}));
+      this->AssertFillNullArray(FillNullForward, value_array, result_array);
+    }
+  }
+}
+
+TYPED_TEST(TestFillNullNumeric, FillNullBackwardLargeInput) {
+  using CType = typename TypeTraits<TypeParam>::CType;
+  random::RandomArrayGenerator rand(/*seed=*/1000);
+  int64_t len_null = 500;
+  int64_t len_random = 1000;
+  std::shared_ptr<Array> array_random =
+      rand.Numeric<TypeParam>(len_random, /*min=*/0, /*max=*/200, /*null_count=*/0);
+
+  if (array_random) {
```

Comment:
Why the check here?

Reply:
Do you mind using `ArrayOf` instead then? It should handle all types: https://github.com/apache/arrow/blob/c6143a2396058dcc31506050238dc0f932aae9ba/cpp/src/arrow/testing/random.h#L380

Reply:
It work for generate a random_array, but currently there is not support for DateTypes for generate a ConstantArray  which is needed for this test.
`
auto constant_array = ConstantArrayGenerator::Numeric<TypeParam>(len_null, x_ptr[len_random - 1])
Concatenate({array_random, constant_array, array_random}))
`

Reply:
Then please document why the check is needed.

Reply:
Or actually, frankly: just add those cases to ConstantArrayGenerator::Numeric, e.g. 

```cpp
case Date32Type:
  return Int32(size, ...)->View(date32());
```

https://github.com/apache/arrow/blob/91e3ac53e2e21736ce6291d73fc37da6fa21259d/cpp/src/arrow/testing/generator.h#L129-L157

Reply:
@lidavidm the types were added and the condition was removed